### PR TITLE
force rewrite memory tables when outdated log is not compact on time

### PR DESCRIPTION
**Pull request by glorv**
ref tikv/tikv#12011

### Problem Summary
In the current implementation, when raft-engine detect outdated log entries that will block log file purge, it will report these regions to tikv and expect tikv to do force compact. Since tikv only do force compact on region leader and ignores request on follower. There are cases that the raft log in leader store is rewrite by tikv but not on one or more follower. Then when the log on follower need to be compact, tikv will ignore the requests because it expect region leader to do so. If there are no newer log be written into this region, as leader's log is already rewritten, it will not trigger log purge. Thus, the old log on the follower will block log file purge.

### Changes
Add a HashMap to record each regions live generation after each purge phase. If a region lives after three times of force compact, raft engine will force rewrite this region to avoid blocking log file purges.

### CODE REVIEW

File: src/purge.rs [hunk 1]
```diff
@@ -23,6 +24,7 @@ const REWRITE_RATIO: f64 = 0.7;
 // Only rewrite region with stale logs less than this threshold.
 const MAX_REWRITE_ENTRIES_PER_REGION: usize = 32;
 const MAX_REWRITE_BATCH_BYTES: usize = 128 * 1024;
+const MAX_COUNT_BEFORE_FORCE_WRITE: u32 = 3;
```
Comment (tabokie) [inline src/purge.rs:27]:
```suggestion
const MAX_COUNT_BEFORE_FORCE_REWRITE: u32 = 3;
```

File: src/purge.rs [hunk 2]
```diff
@@ -35,7 +37,11 @@ where
     listeners: Vec<Arc<dyn EventListener>>,
 
     // Only one thread can run `purge_expired_files` at a time.
-    purge_mutex: Arc<Mutex<()>>,
+    //
+    // the hashmap(raft_group --> count) records raft groups that should be force compact.
+    // if target reft_group are still in the oldest 20% logs after `MAX_COUNT_BEFORE_FORCE_WRITE`
+    // times purge, it will be force rewrite so the logs can be purged.
+    pending_regions: Arc<Mutex<HashMap<u64, u32>>>,
```
Comment (tabokie) [inline src/purge.rs:44]: How about `force_rewrite_candidates`?

File: src/purge.rs [hunk 3]
```diff
@@ -35,7 +37,11 @@ where
     listeners: Vec<Arc<dyn EventListener>>,
 
     // Only one thread can run `purge_expired_files` at a time.
-    purge_mutex: Arc<Mutex<()>>,
+    //
+    // the hashmap(raft_group --> count) records raft groups that should be force compact.
+    // if target reft_group are still in the oldest 20% logs after `MAX_COUNT_BEFORE_FORCE_WRITE`
+    // times purge, it will be force rewrite so the logs can be purged.
```
Comment (tabokie) [inline src/purge.rs:43]:
```suggestion
    // This table records Raft Groups that should be force compacted before. Those that are not compacted in time (after `MAX_EPOCH_BEFORE_FORCE_REWRITE` epochs) will be force rewritten.
```

**Review by tabokie — COMMENTED** [review]
Two more things:
- Adjust the logic in `rewrite_rewrite_queue`: make Raft Groups with more than `MAX_REWRITE_ENTRIES_PER_REGION` rewrite entries be force compacted as well.
- Add a unit test in "engine.rs" to make sure laggers are indeed rewritten.

Oh, and you need to signed off your commit. Maybe `git commit --amend --signoff && git push -f`.

File: src/purge.rs (hunk 1 above)
Comment (tabokie) [reply src/purge.rs:27]: Let's make it 9. 2m window is good enough.

File: src/engine.rs [hunk 4]
```diff
@@ -966,6 +966,90 @@ mod tests {
...
+        // write some small entries to trigger purge, some of the should trigger rewrite
+        // but not compact.
+        for rid in 21..100 {
+            engine.append(rid, 1, 2, Some(&data));
+        }
+
+        let check_purge = |pending_regions: Vec<u64>| {
+            let res = engine.purge_expired_files();
+            assert!(res.is_ok());
+            let mut compact_regions = res.unwrap();
```
Comment (tabokie) [inline src/engine.rs:1007]:
```suggestion
            let mut compact_regions = engine.purge_expired_files().unwrap();
```

File: src/engine.rs [hunk 5]
```diff
@@ -966,6 +966,90 @@ mod tests {
...
+        // purge 3rd.
+        check_purge(vec![1, 2]);
+        let span3 = engine.file_span(LogQueue::Append);
+        assert_eq!(span3, span);
+
+        // pruge the 4th time, it will trigger force rewrite and thus the result should
+        // be empty, and some log files should be purged.
+        check_purge(vec![]);
+        let span4 = engine.file_span(LogQueue::Append);
+        assert!(span4.0 > span.0 + 5);
```
Comment (tabokie) [inline src/engine.rs:1037]: This is a little verbose. Could you use a for loop that break when `purge_expired_files` returns empty list, and panic when the loop ends without encountering an empty list.

File: src/engine.rs [hunk 6]
```diff
@@ -966,6 +966,90 @@ mod tests {
...
+
+        let engine =
+            RaftLogEngine::open_with_file_system(cfg, Arc::new(ObfuscatedFileSystem::new()))
+                .unwrap();
+        let data = vec![b'x'; 1024];
+        // write 50 small entries into region 1~3, it should trigger force compact
+        // report.
+        for rid in 1..4 {
+            for index in 0..50 {
+                engine.append(rid, index, index + 1, Some(&data[..20]));
```
Comment (tabokie) [inline src/engine.rs:990]: why use small entry?

File: src/engine.rs [hunk 7]
```diff
@@ -966,6 +966,90 @@ mod tests {
...
+        for rid in 4..20 {
+            for index in 0..25 {
+                engine.append(rid, index, index + 1, Some(&data[..50]));
+            }
+        }
+        // write some small entries to trigger purge, some of the should trigger rewrite
+        // but not compact.
+        for rid in 21..100 {
+            engine.append(rid, 1, 2, Some(&data));
+        }
```
Comment (tabokie) [inline src/engine.rs:1002]: Maybe we can get the work done with only one type of region writes.

File: src/purge.rs [hunk 8]
```diff
@@ -23,6 +24,7 @@ const REWRITE_RATIO: f64 = 0.7;
 // Only rewrite region with stale logs less than this threshold.
 const MAX_REWRITE_ENTRIES_PER_REGION: usize = 32;
 const MAX_REWRITE_BATCH_BYTES: usize = 128 * 1024;
+const MAX_COUNT_BEFORE_FORCE_REWRITE: u32 = 3;
```
Comment (tabokie) [inline src/purge.rs:27]:
```suggestion
const MAX_COUNT_BEFORE_FORCE_REWRITE: u32 = 9;
```

File: src/engine.rs (hunk 6 above)
Comment (glorv) [reply src/engine.rs:990]: sparse region will trigger rewrite instead of compact.

Comment (tabokie) [reply src/engine.rs:990]: But `sparse` is determined by count of entries, not size of entries.

**Comment by tabokie** [comment]
@glorv The last commit is not signed.

File: src/engine.rs (hunk 6 above)
Comment (glorv) [reply src/engine.rs]: Yes. small entries is used for the convenience of triggering compact and rewrite as the total log size is smaller. I want each region takes about one log file.

**Pull request merged**